File: Python/lld-examples/easy/atm-system/account.py
from datetime import datetime
from typing import List
from transaction import Transaction, TransactionType

class Account:
    """
    Account class representing a bank account
    """
    
    def __init__(self, account_number: str, pin: str, initial_balance: float = 0.0):
        self.account_number = account_number
        self.pin = pin
        self.balance = initial_balance
        self.is_active = True
        self.transactions: List[Transaction] = []
        self.created_at = datetime.now()
        
        # Add initial balance transaction
        if initial_balance > 0:
            self.add_transaction(Transaction(TransactionType.DEPOSIT, initial_balance, "Initial deposit"))
    
    def validate_pin(self, pin: str) -> bool:
        """Validate PIN"""
        return self.pin == pin
    
    def set_pin(self, new_pin: str):
        """Set new PIN"""
        self.pin = new_pin
        self.add_transaction(Transaction(TransactionType.PIN_CHANGE, 0, "PIN changed"))
    
    def withdraw(self, amount: float) -> bool:
        """Withdraw amount from account"""
        if not self.is_active:
            return False
        
        if amount <= 0:
            return False
        
        if self.balance >= amount:
            self.balance -= amount
            self.add_transaction(Transaction(TransactionType.WITHDRAWAL, amount, "Cash withdrawal"))
            return True
        
        return False
    
    def deposit(self, amount: float) -> bool:
        """Deposit amount to account"""
        if not self.is_active:
            return False
        
        if amount <= 0:
            return False
        
        self.balance += amount
        self.add_transaction(Transaction(TransactionType.DEPOSIT, amount, "Cash deposit"))
        return True
    
    def add_transaction(self, transaction: Transaction):
        """Add transaction to history"""
        self.transactions.append(transaction)
    
    def get_transaction_history(self) -> List[Transaction]:
        """Get transaction history"""
        return self.transactions.copy()
    
    def get_balance(self) -> float:
        """Get current balance"""
        return self.balance
    
    def deactivate(self):
        """Deactivate account"""
        self.is_active = False
    
    def activate(self):
        """Activate account"""
        self.is_active = True
    
    def __str__(self):
        return f"Account {self.account_number}: Balance ${self.balance:.2f} ({'Active' if self.is_active else 'Inactive'})" 

File: JAVA/lld-examples/easy/atm-system/src/main/java/com/systemdesign/lld/atm/Account.java
package com.systemdesign.lld.atm;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Account class representing a bank account
 */
public class Account {
    private String accountNumber;
    private String pin;
    private double balance;
    private List<Transaction> transactionHistory;
    private LocalDateTime createdDate;
    private boolean isActive;
    
    public Account(String accountNumber, String pin, double initialBalance) {
        this.accountNumber = accountNumber;
        this.pin = pin;
        this.balance = initialBalance;
        this.transactionHistory = new ArrayList<>();
        this.createdDate = LocalDateTime.now();
        this.isActive = true;
        
        // Add initial balance transaction
        addTransaction(new Transaction(TransactionType.DEPOSIT, initialBalance, "Initial deposit"));
    }
    
    public String getAccountNumber() {
        return accountNumber;
    }
    
    public boolean validatePin(String inputPin) {
        return this.pin.equals(inputPin);
    }
    
    public double getBalance() {
        return balance;
    }
    
    public boolean isActive() {
        return isActive;
    }
    
    public void setActive(boolean active) {
        this.isActive = active;
    }
    
    public boolean withdraw(double amount) {
        if (amount <= 0) {
            return false;
        }
        
        if (balance >= amount) {
            balance -= amount;
            addTransaction(new Transaction(TransactionType.WITHDRAWAL, amount, "ATM withdrawal"));
            return true;
        }
        return false;
    }
    
    public boolean deposit(double amount) {
        if (amount <= 0) {
            return false;
        }
        
        balance += amount;
        addTransaction(new Transaction(TransactionType.DEPOSIT, amount, "ATM deposit"));
        return true;
    }
    
    public boolean changePin(String currentPin, String newPin) {
        if (validatePin(currentPin)) {
            this.pin = newPin;
            addTransaction(new Transaction(TransactionType.PIN_CHANGE, 0, "PIN changed"));
            return true;
        }
        return false;
    }
    
    public void addTransaction(Transaction transaction) {
        transactionHistory.add(transaction);
    }
    
    public List<Transaction> getTransactionHistory() {
        return new ArrayList<>(transactionHistory);
    }
    
    public List<Transaction> getRecentTransactions(int count) {
        List<Transaction> allTransactions = getTransactionHistory();
        int size = allTransactions.size();
        int startIndex = Math.max(0, size - count);
        return allTransactions.subList(startIndex, size);
    }
    
    public LocalDateTime getCreatedDate() {
        return createdDate;
    }
    
    @Override
    public String toString() {
        return "Account{" +
                "accountNumber='" + accountNumber + '\'' +
                ", balance=" + balance +
                ", isActive=" + isActive +
                ", transactionCount=" + transactionHistory.size() +
                '}';
    }
} 